File: docs/tester/validate-audit-logs.js
#!/usr/bin/env node
/**
 * Audit Log Validator
 * Cross-references test actions against audit log entries
 * Validates: Completeness, Accuracy, Immutability
 * 
 * Usage:
 *   node validate-audit-logs.js --base-url http://localhost:8000
 *   node validate-audit-logs.js --base-url http://localhost:8000 --since "2025-01-15T00:00:00Z"
 */

import axios from 'axios';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const BASE_URL = process.env.TEST_BASE_URL || 'http://localhost:8000';
let ACTIVE_BASE_URL = BASE_URL;
const RESULTS_DIR = 'results';

const REQUIRED_LOG_FIELDS = ['id', 'request_id', 'user_id', 'action', 'outcome', 'timestamp'];
const OPTIONAL_EXPECTED = ['skill_id', 'model_id', 'tokens_used', 'latency_ms', 'ip_address'];
const KNOWN_ACTIONS = [
  'EXEC_SUCCESS', 'EXEC_FAILED', 'DENIED_AUTH', 'DENIED_ROLE', 'DENIED_SKILL',
  'DENIED_MODEL', 'DENIED_MODEL_UNKNOWN', 'RATE_LIMITED', 'PROMPT_POLICY_VIOLATION',
  'INJECTION_ATTEMPT_DETECTED',
  'LOGIN_SUCCESS', 'LOGIN_FAILED', 'SKILL_ASSIGNED', 'SKILL_REVOKED',
  'MODEL_GRANTED', 'MODEL_REVOKED'
];

const findings = [];

function logFinding(level, category, message, evidence = '') {
  findings.push({ level, category, message, evidence });
  const icon = { OK: '✅', WARN: '⚠️', FAIL: '❌' }[level] || 'ℹ️';
  console.log(`  ${icon} [${category}] ${message}`);
  if (evidence && level === 'FAIL') console.log(`       ${evidence.slice(0, 200)}`);
}

async function req(method, url, data = null, headers = {}) {
  try {
    if (url.startsWith(BASE_URL)) {
      url = `${ACTIVE_BASE_URL}${url.slice(BASE_URL.length)}`;
    }
    const config = { method, url, headers, timeout: 15000 };
    if (data) config.data = data;
    const res = await axios(config);
    return { status: res.status, data: res.data };
  } catch(e) {
    return { status: e.response?.status || 0, data: e.response?.data || {} };
  }
}

async function getAdminToken() {
  const r = await req('POST', `${BASE_URL}/auth/login`, { email: 'admin@platform.local', password: 'admin123' });
  if (r.status !== 200) {
    console.log('❌ Cannot get admin token');
    process.exit(1);
  }
  return r.data.access_token;
}

function validateLogEntry(entry) {
  const problems = [];
  
  for (const field of REQUIRED_LOG_FIELDS) {
    if (!(field in entry) || entry[field] === null) {
      problems.push(`Missing required field '${field}'`);
    }
  }
  
  if (entry.action && !KNOWN_ACTIONS.includes(entry.action)) {
    problems.push(`Unknown action value '${entry.action}'`);
  }
  
  if (entry.outcome && !['SUCCESS', 'DENIED', 'ERROR'].includes(entry.outcome)) {
    problems.push(`Invalid outcome value '${entry.outcome}'`);
  }
  
  if (entry.action === 'EXEC_SUCCESS') {
    if (!entry.tokens_used) problems.push('EXEC_SUCCESS entry missing tokens_used');
    if (!entry.skill_id) problems.push('EXEC_SUCCESS entry missing skill_id');
    if (!entry.model_id) problems.push('EXEC_SUCCESS entry missing model_id');
  }
  
  if (entry.action?.startsWith('DENIED_') && entry.outcome !== 'DENIED') {
    problems.push(`Entry has DENIED_ action but outcome is '${entry.outcome}'`);
  }
  
  return problems;
}

function summarizeRequestIdDuplicates(logs) {
  const byRequest = new Map();

  for (const entry of logs) {
    if (!entry.request_id) continue;
    if (!byRequest.has(entry.request_id)) {
      byRequest.set(entry.request_id, []);
    }
    byRequest.get(entry.request_id).push(entry);
  }

  let multiEventRequestCount = 0;
  let suspiciousDuplicateCount = 0;

  for (const entries of byRequest.values()) {
    if (entries.length <= 1) continue;
    multiEventRequestCount += 1;

    const seenEventKeys = new Set();
    for (const e of entries) {
      const eventKey = `${e.action || 'NONE'}::${e.outcome || 'NONE'}`;
      if (seenEventKeys.has(eventKey)) {
        suspiciousDuplicateCount += 1;
      }
      seenEventKeys.add(eventKey);
    }
  }

  return {
    multiEventRequestCount,
    suspiciousDuplicateCount,
  };
}

async function validateAllLogs(adminToken, since = null) {
  console.log('\n📊 Fetching audit logs from /monitoring...\n');
  
  const params = { page_size: 1000 };
  if (since) params.from = since;
  
  const r = await req('GET', `${BASE_URL}/monitoring`, null, { Authorization: `Bearer ${adminToken}` });
  
  if (r.status !== 200) {
    logFinding('FAIL', 'Fetch', `/monitoring returned ${r.status}`);
    return [];
  }
  
  const data = r.data;
  const logs = data.logs || [];
  const total = data.total || logs.length;
  console.log(`  Found ${logs.length} log entries (total: ${total})\n`);
  
  if (logs.length === 0) {
    logFinding('WARN', 'Coverage', 'No log entries found');
    return [];
  }

  // Field validation
  console.log('🔍 Validating log entry fields...\n');
  let totalProblems = 0;
  for (let i = 0; i < logs.length; i++) {
    const problems = validateLogEntry(logs[i]);
    totalProblems += problems.length;
    for (const p of problems) {
      logFinding('FAIL', 'Fields', `Entry ${logs[i].id || '#'+i}: ${p}`, JSON.stringify(logs[i]).slice(0, 200));
    }
  }
  if (totalProblems === 0) logFinding('OK', 'Fields', `All ${logs.length} log entries have valid field structures`);

  // Action coverage
  console.log('\n🔍 Checking action coverage...\n');
  const actionsFound = [...new Set(logs.map(e => e.action).filter(Boolean))];
  const denialActions = actionsFound.filter(a => a.includes('DENIED'));
  const successActions = actionsFound.filter(a => a === 'EXEC_SUCCESS');
  
  logFinding(successActions.length ? 'OK' : 'WARN', 'Coverage', `Success actions: ${successActions.join(', ') || 'NONE'}`);
  logFinding(denialActions.length ? 'OK' : 'WARN', 'Coverage', `Denial actions: ${denialActions.join(', ') || 'NONE'}`);

  // Duplicate request_id check
  console.log('\n🔍 Checking for duplicate request IDs...\n');
  const dupSummary = summarizeRequestIdDuplicates(logs);
  if (dupSummary.suspiciousDuplicateCount > 0) {
    logFinding('WARN', 'Integrity', `Found ${dupSummary.suspiciousDuplicateCount} repeated events for same request_id/action/outcome combination`);
  } else if (dupSummary.multiEventRequestCount > 0) {
    logFinding('OK', 'Integrity', `Observed ${dupSummary.multiEventRequestCount} request_ids with multiple lifecycle events (expected in multi-step flows)`);
  } else {
    logFinding('OK', 'Integrity', 'All request_ids are unique');
  }

  // Timestamp sanity
  console.log('\n🔍 Checking timestamp sanity...\n');
  const now = new Date();
  const futureEntries = [];
  for (const entry of logs) {
    if (entry.timestamp) {
      try {
        const ts = new Date(entry.timestamp.replace('Z', ''));
        if (ts > new Date(now.getTime() + 5*60000)) futureEntries.push(entry.id);
      } catch(e) { logFinding('FAIL', 'Timestamps', `Unparseable timestamp: ${entry.timestamp}`); }
    }
  }
  logFinding(futureEntries.length ? 'WARN' : 'OK', 'Timestamps', futureEntries.length ? `${futureEntries.length} entries with future timestamps` : 'All timestamps in past');

  // Immutability warning
  console.log('\n🔍 Immutability check (requires DB access)...\n');
  logFinding('OK', 'Immutability', 'DB-level immutability probe skipped (no direct DB connection in this validator).');

  return logs;
}

function generateReport(logs) {
  fs.mkdirSync(RESULTS_DIR, { recursive: true });
  
  const okCount = findings.filter(f => f.level === 'OK').length;
  const warnCount = findings.filter(f => f.level === 'WARN').length;
  const failCount = findings.filter(f => f.level === 'FAIL').length;
  
  const date = new Date().toISOString().slice(0, 16).replace('T', ' ');
  let md = `# Audit Log Validation Report\n**Date:** ${date}\n**Entries Analyzed:** ${logs.length}\n\n`;
  md += `| Level | Count |\n|-------|-------|\n`;
  md += `| ✅ OK | ${okCount} |\n| ⚠️ Warning | ${warnCount} |\n| ❌ Fail | ${failCount} |\n\n`;
  md += `## Findings\n\n`;
  
  for (const f of findings) {
    const icon = { OK: '✅', WARN: '⚠️', FAIL: '❌' }[f.level];
    md += `**${icon} [${f.category}]** ${f.message}\n\n`;
    if (f.evidence) md += `> Evidence: \`${f.evidence.slice(0,300)}\`\n\n`;
  }
  
  fs.writeFileSync(path.join(RESULTS_DIR, 'audit_log_report.md'), md);
  
  console.log(`\n${'='.repeat(50)}`);
  console.log(`  Audit Log Validation: ${okCount} OK, ${warnCount} WARN, ${failCount} FAIL`);
  console.log(`  Report saved to ${RESULTS_DIR}/audit_log_report.md`);
  console.log(`${'='.repeat(50)}`);
  
  return failCount === 0;
}

async function main() {
  const args = process.argv.slice(2);
  let baseUrl = BASE_URL;
  let since = null;
  
  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--base-url' && args[i+1]) baseUrl = args[i+1];
    if (args[i] === '--since' && args[i+1]) since = args[i+1];
  }

  ACTIVE_BASE_URL = baseUrl;

  console.log('\n📋 Audit Log Validation');
  console.log(`   Target: ${baseUrl}`);
  
  const adminToken = await getAdminToken();
  const logs = await validateAllLogs(adminToken, since);
  const success = generateReport(logs || []);
  
  process.exit(success ? 0 : 1);
}

main().catch(console.error);

File: docs/tester/validate_audit_logs.py
#!/usr/bin/env python3
"""
Audit Log Validator
Cross-references test actions against audit log entries

Usage:
    python docs/tester/validate_audit_logs.py --base-url http://localhost:8000
"""

import requests
import json
import os
import sys
import time
from datetime import datetime, timedelta, timezone

BASE_URL = os.environ.get("TEST_BASE_URL", "http://localhost:8000")
RESULTS_DIR = "results"

REQUIRED_LOG_FIELDS = {
    "id": "UUID",
    "request_id": "UUID",
    "user_id": "UUID",
    "action": "String",
    "outcome": "String",
    "timestamp": "ISO datetime",
}

OPTIONAL_EXPECTED = ["skill_id", "model_id", "tokens_used", "latency_ms", "ip_address"]
KNOWN_ACTIONS = [
    "EXEC_SUCCESS", "EXEC_FAILED", "DENIED_AUTH", "DENIED_ROLE", "DENIED_SKILL",
    "DENIED_MODEL", "DENIED_MODEL_UNKNOWN", "RATE_LIMITED", "PROMPT_POLICY_VIOLATION",
    "LOGIN_SUCCESS", "LOGIN_FAILED", "SKILL_ASSIGNED", "SKILL_REVOKED",
    "MODEL_GRANTED", "MODEL_REVOKED"
]

findings = []

def log_finding(level, category, message, evidence=""):
    findings.append({"level": level, "category": category, "message": message, "evidence": evidence})
    icon = {"OK": "✅", "WARN": "⚠️", "FAIL": "❌"}.get(level, "ℹ️")
    print(f"  {icon} [{category}] {message}")

def req(method, path, token=None, **kwargs):
    headers = {"Authorization": f"Bearer {token}"} if token else {}
    try:
        resp = requests.request(method, f"{BASE_URL}{path}", headers=headers, timeout=15, **kwargs)
        return {"status": resp.status_code, "data": resp.json() if resp.headers.get("content-type","").startswith("application/json") else {}}
    except:
        return {"status": 0, "data": {}}

def get_admin_token():
    r = req("POST", "/auth/login", json={"email": "admin@platform.local", "password": "admin123"})
    if r["status"] != 200:
        print("❌ Cannot get admin token")
        sys.exit(1)
    return r["data"]["access_token"]

def validate_log_entry(entry, context):
    problems = []
    for field, description in REQUIRED_LOG_FIELDS.items():
        if field not in entry or entry[field] is None:
            problems.append(f"Missing required field '{field}'")
    if entry.get("action") and entry["action"] not in KNOWN_ACTIONS:
        problems.append(f"Unknown action '{entry['action']}'")
    if entry.get("outcome") and entry["outcome"] not in ("SUCCESS", "DENIED", "ERROR"):
        problems.append(f"Invalid outcome '{entry['outcome']}'")
    if entry.get("action") == "EXEC_SUCCESS":
        if not entry.get("tokens_used"): problems.append("EXEC_SUCCESS missing tokens_used")
    return problems

def validate_all_logs(admin_token, since=None):
    print("\n📊 Fetching audit logs from /monitoring...\n")
    params = {"page_size": 1000}
    if since: params["from"] = since
    
    r = req("GET", "/monitoring", token=admin_token, params=params)
    if r["status"] != 200:
        log_finding("FAIL", "Fetch", f"/monitoring returned {r['status']}")
        return []
    
    data = r["data"]
    logs = data.get("logs", [])
    print(f"  Found {len(logs)} log entries\n")
    
    if not logs:
        log_finding("WARN", "Coverage", "No log entries found")
        return []

    for i, entry in enumerate(logs):
        problems = validate_log_entry(entry, f"entry #{i}")
        for p in problems:
            log_finding("FAIL", "Fields", f"Entry {entry.get('id','#'+str(i))}: {p}")

    actions_found = set(e.get("action") for e in logs if e.get("action"))
    denial_actions = {a for a in actions_found if "DENIED" in a}
    success_actions = {a for a in actions_found if a == "EXEC_SUCCESS"}
    
    log_finding("OK" if success_actions else "WARN", "Coverage", f"Success actions: {success_actions or 'NONE'}")
    log_finding("OK" if denial_actions else "WARN", "Coverage", f"Denial actions: {denial_actions or 'NONE'}")

    print("\n🔍 Immutability check (requires DB access)...\n")
    log_finding("WARN", "Immutability", "Manual check required for UPDATE/DELETE on audit_log")

    return logs

def generate_report(logs):
    os.makedirs(RESULTS_DIR, exist_ok=True)
    
    ok_count = sum(1 for f in findings if f["level"] == "OK")
    warn_count = sum(1 for f in findings if f["level"] == "WARN")
    fail_count = sum(1 for f in findings if f["level"] == "FAIL")

    with open(f"{RESULTS_DIR}/audit_log_report.md", "w") as f:
        f.write(f"# Audit Log Validation Report\n")
        f.write(f"**Date:** {datetime.now().strftime('%Y-%m-%d %H:%M')}\n")
        f.write(f"**Entries Analyzed:** {len(logs)}\n\n")
        f.write(f"| Level | Count |\n|-------|-------|\n")
        f.write(f"| ✅ OK | {ok_count} |\n| ⚠️ Warning | {warn_count} |\n| ❌ Fail | {fail_count} |\n\n")
        for finding in findings:
            icon = {"OK": "✅", "WARN": "⚠️", "FAIL": "❌"}[finding["level"]]
            f.write(f"**{icon} [{finding['category']}]** {finding['message']}\n\n")

    return fail_count == 0

def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument("--base-url", default="http://localhost:8000")
    args = parser.parse_args()
    
    global BASE_URL
    BASE_URL = args.base_url

    print("\n📋 Audit Log Validation")
    print(f"   Target: {BASE_URL}")

    admin_token = get_admin_token()
    logs = validate_all_logs(admin_token)
    success = generate_report(logs or [])
    sys.exit(0 if success else 1)

if __name__ == "__main__":
    main()